### ISSUE TITLE
kv: NotLeaderError returned from dist_sender instead of being retried 

### ISSUE THREAD

**Issue by tschottdorf** [cockroachdb/cockroach#6209]
The following test appears to have failed:

[#16639](https://circleci.com/gh/cockroachdb/cockroach/16639):

```
I160421 18:37:23.044454 acceptance/dynamic_client.go:61  closing connection to 127.0.0.1:32863
I160421 18:37:23.047146 acceptance/dynamic_client.go:61  closing connection to 127.0.0.1:32865
I160421 18:37:23.147714 acceptance/cluster/localcluster.go:654  stopping
I160421 18:37:23.147982 acceptance/cluster/localcluster.go:552  event stream done, resetting...: net/http: request canceled
I160421 18:37:23.148002 acceptance/cluster/localcluster.go:575  events monitor exits
--- FAIL: TestRepair (10.62s)
	load.go:31: pq: range 1: replica <nil> not leader; leader is node_id:1 store_id:1 replica_id:1 
=== RUN   TestDockerRuby
I160421 18:37:25.557358 acceptance/cluster/localcluster.go:289  Initializing Cluster ruby:
{"name":"ruby","nodes":[{"count":1,"stores":[{"count":1,"max_ranges":0}]}],"duration":5000000000}
I160421 18:37:25.559240 acceptance/cluster/docker.go:105  ImagePull cockroachdb/builder:20160305-182433 already exists
I160421 18:37:26.147079 acceptance/cluster/localcluster.go:586  creating certs (1024bit) in: /home/ubuntu/cockroach/.localcluster.certs.030958138
I160421 18:37:26.251498 acceptance/cluster/localcluster.go:537  events monitor starts
I160421 18:37:26.977623 acceptance/cluster/localcluster.go:485  *** started roach0 ***
  ui:        https://127.0.0.1:32868
  trace:     https://127.0.0.1:32868/debug/requests
--
--- SKIP: TestBuildBabyCluster (0.00s)
	terraform_test.go:30: only enabled during testing
=== RUN   TestFiveNodesAndWriters
--- SKIP: TestFiveNodesAndWriters (0.00s)
	util_test.go:68: running in docker mode
FAIL
ok github.com/cockroachdb/cockroach/acceptance 1337s

```
Please assign, take a look and update the issue accordingly.

**Comment by tschottdorf** [comment]
Again in `TestNodeRestart`: https://circle-artifacts.com/gh/cockroachdb/cockroach/17294/artifacts/0/tmp/circle-artifacts.XEhe6Sr/acceptance_deploy.log

**Comment by bdarnell** [comment]
This will probably be fixed by my work on #2500

**Comment by tschottdorf** [comment]
Let's close this, Ben has done a lot of work in this code path.
